tasks/message/watch_eth.go

```diff
@@ -63,110 +92,169 @@ func (mw *MessageWatcherEth) update() {
 	ctx := context.Background()
 
 	bestBlockNumber := mw.bestBlockNumber.Load()
+	log.Debugw("MessageWatcherEth update starting", "bestBlockNumber", bestBlockNumber)
 
 	confirmedBlockNumber := new(big.Int).Sub(bestBlockNumber, big.NewInt(MinConfidence))
 	if confirmedBlockNumber.Sign() < 0 {
 		// Not enough blocks yet
+		log.Debugw("Not enough blocks for confirmations", "bestBlockNumber", bestBlockNumber, "minConfidence", MinConfidence)
 		return
 	}
 
 	machineID := mw.ht.ResourcesAvailable().MachineID
 
 	// Assign pending transactions with null owner to ourselves
 	{
-		n, err := mw.db.Exec(ctx, `UPDATE message_waits_eth SET waiter_machine_id = $1 WHERE waiter_machine_id IS NULL AND tx_status = 'pending'`, machineID)
+		n, err := mw.txMgr.AssignPendingToMachine(ctx, int64(machineID))
 		if err != nil {
 			log.Errorf("failed to assign pending transactions: %+v", err)
 			return
 		}
 		if n > 0 {
-			log.Debugw("assigned pending transactions to ourselves", "assigned", n)
+			log.Infow("Assigned orphaned pending transactions", "count", n, "machineID", machineID)
 		}
 	}
 
 	// Get transactions assigned to us
-	var txs []struct {
-		TxHash string `db:"signed_tx_hash"`
-	}
-
-	err := mw.db.Select(ctx, &txs, `SELECT signed_tx_hash FROM message_waits_eth WHERE waiter_machine_id = $1 AND tx_status = 'pending' LIMIT 10000`, machineID)
+	txHashes, err := mw.txMgr.GetPendingForMachine(ctx, int64(machineID))
 	if err != nil {
 		log.Errorf("failed to get assigned transactions: %+v", err)
 		return
 	}
 
+	log.Infow("Processing pending transactions", "count", len(txHashes), "machineID", machineID)
+
+	// Track statistics
+	var processed, stillPending, waitingConfirmations, confirmed, errorCount int
+
 	// Check if any of the transactions we have assigned are now confirmed
-	for _, tx := range txs {
-		txHash := common.HexToHash(tx.TxHash)
+	for _, txHashStr := range txHashes {
+		txHash := common.HexToHash(txHashStr)
+		log.Debugw("Checking transaction", "txHash", txHash.Hex())
+
+		ethCtx, ethCancel := context.WithTimeout(ctx, mw.ethCallTimeout)
 
-		receipt, err := mw.api.TransactionReceipt(ctx, txHash)
+		receipt, err := mw.api.TransactionReceipt(ethCtx, txHash)
```

Comment:
Is the intention here, that when the backend of this api (lotus) fails to "do the thing" for whatever reason, that this task is retried later by the task engine? Rather than implement the retry logic here directly? i.e. the task engine is the retry logic?

Reply:
Yes, this is dealing with an actual bug we've experienced - probably caused by a mismatch between how Lotus expresses an error and what go-ethereum's ethclient expects as an error (still don't understand the nuances). So ethclient just hangs indefinately, locking up the whole transaction watcher process. So this timeout would bail, the transaction would be left in a `pending` state and would be retried in the next loop, so nothing is lost and it safely retries next chance.